libraries/AP_Compass/AP_Compass_BMM150.cpp

```diff
@@ -0,0 +1,279 @@
+    int32_t val;
+
+    if (!rhall) {
+        rhall = _dig_xyz1;
+    }
+
+    if (is_x) {
+        txy1 = _dig_x1;
+        txy2 = _dig_x2;
+    } else {
+        txy1 = _dig_y1;
+        txy2 = _dig_y2;
+    }
+
+    /* TODO: make compensation calculations works on big-endian CPUs */
+    val = ((int16_t)(((uint16_t)((((int32_t)_dig_xyz1) << 14) / rhall)) -
+          ((uint16_t)0x4000)));
+    val = ((int16_t)((((int32_t)xy) * ((((((((int32_t)_dig_xy2) *
+          ((((int32_t)val) * ((int32_t)val)) >> 7)) + (((int32_t)val) *
+          ((int32_t)(((int16_t)_dig_xy1) << 7)))) >> 9) +
+          ((int32_t)0x100000)) * ((int32_t)(((int16_t)txy2) +
+          ((int16_t)0xA0)))) >> 12)) >> 13)) + (((int16_t)txy1) << 3);
+
+    return (int32_t)val;
+
+}
+
+int16_t AP_Compass_BMM150::_compensate_z(int16_t z, uint16_t rhall)
+{
+    int32_t val;
+
+    /* TODO: make compensation calculations works on big-endian CPUs */
+    val = (((((int32_t)(z - _dig_z4)) << 15) - ((((int32_t)_dig_z3) *
+          ((int32_t)(((int16_t)rhall) - ((int16_t)_dig_xyz1)))) >> 2)) /
+          (_dig_z2 + ((int16_t)(((((int32_t)_dig_z1) *
+          ((((int16_t)rhall) << 1))) + (1 << 15)) >> 16))));
+
+    return val;
+}
+
+void AP_Compass_BMM150::_update()
+{
+    const uint32_t time_us = AP_HAL::micros();
+
+    if (time_us - _last_update_timestamp < MICRO_SECONDS_BETWEEN_MEASUREMENTS) {
+        return;
+    }
+
+    if (!_dev->get_semaphore()->take_nonblocking()) {
+        return;
+    }
+
+    uint16_t data[4];
+    _dev->read_registers(DATA_X_LSB_REG, (uint8_t *) &data, sizeof(data));
+    _dev->get_semaphore()->give();
+
+    /* Checking data ready status */
+    if (!data[3] & 0x1) {
+        return;
+    }
+
+    const uint16_t rhall = le16toh(data[3] >> 2);
+    Vector3f raw_field = Vector3f();
+
+    int16_t val = le16toh(((int16_t)data[0]) >> 3);
+    val = _compensate_xy(val, rhall, true);
+    raw_field.x = val;
+
+    val = le16toh(((int16_t)data[1]) >> 3);
+    val = _compensate_xy(val, rhall, false);
+    raw_field.y = val;
+
+    val = le16toh(((int16_t)data[2]) >> 1);
+    val = _compensate_z(val, rhall);
+    raw_field.z = val;
+
+    /* apply sensitivity scale */
+    raw_field /= 16;
+    /* uT to milligaus */
+    raw_field *= 10;
+
+    /* rotate raw_field from sensor frame to body frame */
+    rotate_field(raw_field, _compass_instance);
+
+    /* publish raw_field (uncorrected point sample) for calibration use */
+    publish_raw_field(raw_field, time_us, _compass_instance);
+
+    /* correct raw_field for known errors */
+    correct_field(raw_field, _compass_instance);
+
+    _mag_accum += raw_field;
+    _accum_count++;
+    if (_accum_count == 10) {
+        _mag_accum /= 2;
+        _accum_count = 5;
+    }
+
+    _last_update_timestamp = time_us;
+}
+
+void AP_Compass_BMM150::read()
+{
+    if (_accum_count == 0) {
+        return;
+    }
+
+    hal.scheduler->suspend_timer_procs();
+
+    Vector3f field(_mag_accum);
```

Comment:
Ha? on the line 281 of AK8963 it is just releasing the lock of the bus.
And there is no call to suspend_timer_proc() on the _update() of AK8963.